packages/user/Tokens/query-service/src/lib.rs

```diff
@@ -171,20 +193,86 @@ mod service {
                 .query()
         }
 
+        /// Returns a paginated subset of all historical events (that haven't yet been pruned
+        /// from this node) related to changes to the balances of the specified token for the
+        /// specified user.
+        /// Changes in the balance of at least one user happen when tokens are recalled, burned,
+        /// credited, debited, uncredited, rejected.
         async fn balChanges(
             &self,
             token_id: TID,
             account: AccountNumber,
+            counter_parties: Option<Vec<String>>,
+            excluded_counter_parties: Option<Vec<String>>,
+            action: Option<String>,
+            amount_min: Option<String>,
+            amount_max: Option<String>,
+            memo: Option<String>,
             first: Option<i32>,
             last: Option<i32>,
             before: Option<String>,
             after: Option<String>,
         ) -> async_graphql::Result<Connection<u64, BalanceEvent>> {
+            let precision = TokenTable::with_service(tokens::SERVICE)
+                .get_index_pk()
+                .get(&token_id)
+                .ok_or_else(|| {
+                    async_graphql::Error::new(format!("token '{}' not found", token_id))
+                })?
+                .precision
+                .value();
+
+            // Create fixed-precision string versions of the specified amount bounds
+            let min_fixed = amount_min.as_deref().map(|s| to_fixed(s, precision));
+            let max_fixed = amount_max.as_deref().map(|s| to_fixed(s, precision));
+
+            let mut conditions = vec![
+                format!("token_id = {}", token_id),
+                format!("account = '{}'", account),
+            ];
+
+            if let Some(cps) = &counter_parties {
+                if !cps.is_empty() {
+                    let cp_conditions: Vec<String> = cps
+                        .iter()
+                        .map(|cp| format!("counter_party LIKE '%{}%'", cp))
+                        .collect();
+                    conditions.push(format!("({})", cp_conditions.join(" OR ")));
+                }
+            }
+
+            if let Some(excluded_cps) = &excluded_counter_parties {
+                if !excluded_cps.is_empty() {
+                    let excluded_conditions: Vec<String> = excluded_cps
+                        .iter()
+                        .map(|cp| format!("counter_party NOT LIKE '%{}%'", cp))
+                        .collect();
+                    conditions.push(format!("({})", excluded_conditions.join(" AND ")));
```

Comment:
Are the parentheses needed? I think `excluded_conditions` could just be appended to `conditions`.